File: android/lynx-shell/src/main/kotlin/com/ota/android/sdk/BundleRuntime.kt
package com.ota.android.sdk

import java.io.File
import java.io.IOException

/**
 * Bundle 运行时门面：路由按 `lynxAppId + bundleName` 取已提交 current，绝不读取
 * `.staging`。pageId 解析只保留在 [OtaSdk.currentTemplatePath] 兼容入口中。
 */
class BundleRuntime(private val transaction: ReleaseTransaction) {
  constructor(storageRoot: File) : this(ReleaseTransaction(storageRoot))

  @Throws(IOException::class, OtaSdkException::class)
  fun current(scope: ReleaseTransaction.ReleaseScope, bundleName: String): File? {
    return transaction.currentBundle(scope, bundleName)
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun ensureBundleReady(scope: ReleaseTransaction.ReleaseScope, bundleName: String): File {
    return transaction.ensureBundleReady(scope, bundleName)
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun current(lynxAppId: String, bundleName: String): File? {
    return transaction.currentBundle(lynxAppId, bundleName)
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun ensureBundleReady(lynxAppId: String, bundleName: String): File {
    return transaction.ensureBundleReady(lynxAppId, bundleName)
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun current(scope: ReleaseTransaction.ReleaseScope): OtaModels.InstalledRelease? {
    return transaction.current(scope)
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun currentTemplatePath(scope: ReleaseTransaction.ReleaseScope, pageId: Int): File? {
    val release = transaction.current(scope) ?: return null
    return release.bundles.firstOrNull { it.pageId == pageId }?.let { bundle ->
      transaction.currentBundle(scope, bundle.bundlePath)
    }
  }

  @Throws(IOException::class, OtaSdkException::class)
  fun currentTemplatePath(lynxAppId: String, pageId: Int): File? {
    return transaction.currentTemplatePath(lynxAppId, pageId)
  }
}

File: ios/OtaIOSSDK/Sources/OtaIOSSDK/BundleRuntime.swift
import Foundation

/// iOS Bundle Runtime 门面。
///
/// 路由只通过这个 actor 读取已提交 current，不直接拼接 staged、previous 或
/// release 目录路径。网络 repair 仍由 `OtaSDK.ensureBundleReady` 负责。
public actor BundleRuntime {
    private let transaction: ReleaseTransaction

    public init(transaction: ReleaseTransaction) {
        self.transaction = transaction
    }

    /// 精确按 `hostApp + lynxAppId + bundleName` 读取 current 文件。
    public func current(scope: OtaReleaseScope, bundleName: String) async throws -> URL? {
        try await transaction.currentBundle(scope: scope, bundleName: bundleName)
    }

    public func current(scope: OtaReleaseScope) async -> OtaInstalledRelease? {
        await transaction.current(scope: scope)
    }

    /// 当前 release 无该 bundle 时抛结构化 not-found；不会读取 staged 文件。
    public func ensureBundleReady(scope: OtaReleaseScope, bundleName: String) async throws -> URL {
        guard let url = try await current(scope: scope, bundleName: bundleName) else {
            throw OtaSDKError.bundleNotFound(
                lynxAppId: scope.lynxAppId,
                bundleName: bundleName
            )
        }
        return url
    }
}
